File: ios/MVVMDemo/Services/ValidationService.m
//
//  ValidationService.m
//  MVVMDemo
//
//  Created by MVVMDemo on 2025-06-17.
//

#import "ValidationService.h"
#import "ValidationHelpers.h"
#import "ValidationError.h"

// Error domain and constants
NSString * const ValidationErrorDomain = @"MVVMDemo.ValidationErrorDomain";
NSString * const ValidationErrorCodeKey = @"MVVMDemo.ValidationErrorCodeKey";

@implementation ValidationService

+ (ValidationService *)shared {
    static ValidationService *sharedInstance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedInstance = [[self alloc] init];
    });
    return sharedInstance;
}

// MARK: - ValidationServiceProtocol Implementation

- (BOOL)validateEmail:(NSString *)email error:(NSError **)error {
    if ([ValidationHelpers isEmailEmpty:email]) {
        if (error) {
            *error = [self errorWithValidationError:ValidationErrorEmailEmpty];
        }
        return NO;
    }
    
    if ([ValidationHelpers isEmailTooLong:email]) {
        if (error) {
            *error = [self errorWithValidationError:ValidationErrorEmailTooLong];
        }
        return NO;
    }
    
    if (![ValidationHelpers isValidEmail:email]) {
        if (error) {
            *error = [self errorWithValidationError:ValidationErrorEmailInvalid];
        }
        return NO;
    }
    
    return YES;
}

- (BOOL)validatePassword:(NSString *)password error:(NSError **)error {
    if ([ValidationHelpers isPasswordEmpty:password]) {
        if (error) {
            *error = [self errorWithValidationError:ValidationErrorPasswordEmpty];
        }
        return NO;
    }
    
    if ([ValidationHelpers isPasswordTooShort:password]) {
        if (error) {
            *error = [self errorWithValidationError:ValidationErrorPasswordTooShort];
        }
        return NO;
    }
    
    if ([ValidationHelpers isPasswordTooLong:password]) {
        if (error) {
            *error = [self errorWithValidationError:ValidationErrorPasswordTooLong];
        }
        return NO;
    }
    
    if (![ValidationHelpers hasUppercaseLetter:password]) {
        if (error) {
            *error = [self errorWithValidationError:ValidationErrorPasswordMissingUppercase];
        }
        return NO;
    }
    
    if (![ValidationHelpers hasLowercaseLetter:password]) {
        if (error) {
            *error = [self errorWithValidationError:ValidationErrorPasswordMissingLowercase];
        }
        return NO;
    }
    
    if (![ValidationHelpers hasNumber:password]) {
        if (error) {
            *error = [self errorWithValidationError:ValidationErrorPasswordMissingNumber];
        }
        return NO;
    }
    
    if (![ValidationHelpers hasSpecialCharacter:password]) {
        if (error) {
            *error = [self errorWithValidationError:ValidationErrorPasswordMissingSpecialCharacter];
        }
        return NO;
    }
    
    if (![ValidationHelpers isPasswordStrong:password]) {
        if (error) {
            *error = [self errorWithValidationError:ValidationErrorPasswordWeak];
        }
        return NO;
    }
    
    return YES;
}

- (BOOL)validateName:(NSString *)name error:(NSError **)error {
    if ([ValidationHelpers isNameEmpty:name]) {
        if (error) {
            *error = [self errorWithValidationError:ValidationErrorNameEmpty];
        }
        return NO;
    }
    
    if ([ValidationHelpers isNameTooShort:name]) {
        if (error) {
            *error = [self errorWithValidationError:ValidationErrorNameTooShort];
        }
        return NO;
    }
    
    if ([ValidationHelpers isNameTooLong:name]) {
        if (error) {
            *error = [self errorWithValidationError:ValidationErrorNameTooLong];
        }
        return NO;
    }
    
    if (![ValidationHelpers hasValidNameCharacters:name]) {
        if (error) {
            *error = [self errorWithValidationError:ValidationErrorNameInvalidCharacters];
        }
        return NO;
    }
    
    return YES;
}

- (BOOL)validatePassword:(NSString *)password 
        confirmPassword:(NSString *)confirmPassword 
                  error:(NSError **)error {
    if (![ValidationHelpers stringsMatch:password string2:confirmPassword]) {
        if (error) {
            *error = [self errorWithValidationError:ValidationErrorPasswordMismatch];
        }
        return NO;
    }
    
    return YES;
}

- (NSArray<NSNumber *> *)validateRegistrationCredentials:(RegistrationCredentials *)credentials {
    NSMutableArray<NSNumber *> *errors = [NSMutableArray array];
    
    // Validate email
    NSError *emailError = nil;
    if (![self validateEmail:credentials.email error:&emailError]) {
        [errors addObject:@(emailError.code)];
    }
    
    // Validate password
    NSError *passwordError = nil;
    if (![self validatePassword:credentials.password error:&passwordError]) {
        [errors addObject:@(passwordError.code)];
    }
    
    // Validate password confirmation
    NSError *confirmError = nil;
    if (![self validatePassword:credentials.password 
                confirmPassword:credentials.confirmPassword 
                          error:&confirmError]) {
        [errors addObject:@(confirmError.code)];
    }
    
    // Validate name
    NSError *nameError = nil;
    if (![self validateName:credentials.fullName error:&nameError]) {
        [errors addObject:@(nameError.code)];
    }
    
    return [errors copy];
}

- (NSArray<NSNumber *> *)validateLoginCredentials:(UserCredentials *)credentials {
    NSMutableArray<NSNumber *> *errors = [NSMutableArray array];
    
    // Validate email
    NSError *emailError = nil;
    if (![self validateEmail:credentials.email error:&emailError]) {
        [errors addObject:@(emailError.code)];
    }
    
    // Validate password (basic validation for login)
    if ([ValidationHelpers isPasswordEmpty:credentials.password]) {
        [errors addObject:@(ValidationErrorPasswordEmpty)];
    }
    
    return [errors copy];
}

// MARK: - Convenience Methods

- (NSString *)getEmailErrorMessage:(NSString *)email {
    NSError *error = nil;
    if ([self validateEmail:email error:&error]) {
        return nil;
    }
    return error.localizedDescription;
}

- (NSString *)getPasswordErrorMessage:(NSString *)password {
    NSError *error = nil;
    if ([self validatePassword:password error:&error]) {
        return nil;
    }
    return error.localizedDescription;
}

- (NSString *)getNameErrorMessage:(NSString *)name {
    NSError *error = nil;
    if ([self validateName:name error:&error]) {
        return nil;
    }
    return error.localizedDescription;
}

- (NSString *)getPasswordConfirmationErrorMessage:(NSString *)password 
                                  confirmPassword:(NSString *)confirmPassword {
    NSError *error = nil;
    if ([self validatePassword:password 
                confirmPassword:confirmPassword 
                          error:&error]) {
        return nil;
    }
    return error.localizedDescription;
}

// MARK: - Private Helper Methods

- (NSError *)errorWithValidationError:(ValidationError)validationError {
    NSString *description = [ValidationErrorHelper localizedDescriptionForError:validationError];
    NSInteger code = [ValidationErrorHelper errorCodeForError:validationError];
    
    NSDictionary *userInfo = @{
        NSLocalizedDescriptionKey: description,
        ValidationErrorCodeKey: @(code)
    };
    
    return [NSError errorWithDomain:ValidationErrorDomain code:code userInfo:userInfo];
}

@end

File: ios/MVVMDemo/Services/ValidationService.swift
//
//  ValidationService.swift
//  MVVMDemo
//
//  Created by MVVMDemo on 2025-06-17.
//

import Foundation

/// Concrete implementation of ValidationServiceProtocol
@objc public class ValidationService: NSObject, ValidationServiceProtocol {
    
    // MARK: - Singleton
    
    /// Shared instance for easy access
    @objc public static let shared = ValidationService()
    
    // MARK: - ValidationServiceProtocol Implementation
    
    /// Validate email address
    /// - Parameter email: Email address to validate
    /// - Returns: Validation result with error if invalid
    @objc public func validateEmail(_ email: String) -> Result<Void, ValidationError> {
        if ValidationHelpers.isEmailEmpty(email) {
            return .failure(.emailEmpty)
        }
        
        if ValidationHelpers.isEmailTooLong(email) {
            return .failure(.emailTooLong)
        }
        
        if !ValidationHelpers.isValidEmail(email) {
            return .failure(.emailInvalid)
        }
        
        return .success(())
    }
    
    /// Validate password
    /// - Parameter password: Password to validate
    /// - Returns: Validation result with error if invalid
    @objc public func validatePassword(_ password: String) -> Result<Void, ValidationError> {
        if ValidationHelpers.isPasswordEmpty(password) {
            return .failure(.passwordEmpty)
        }
        
        if ValidationHelpers.isPasswordTooShort(password) {
            return .failure(.passwordTooShort)
        }
        
        if ValidationHelpers.isPasswordTooLong(password) {
            return .failure(.passwordTooLong)
        }
        
        if !ValidationHelpers.hasUppercaseLetter(password) {
            return .failure(.passwordMissingUppercase)
        }
        
        if !ValidationHelpers.hasLowercaseLetter(password) {
            return .failure(.passwordMissingLowercase)
        }
        
        if !ValidationHelpers.hasNumber(password) {
            return .failure(.passwordMissingNumber)
        }
        
        if !ValidationHelpers.hasSpecialCharacter(password) {
            return .failure(.passwordMissingSpecialCharacter)
        }
        
        if !ValidationHelpers.isPasswordStrong(password) {
            return .failure(.passwordWeak)
        }
        
        return .success(())
    }
    
    /// Validate name
    /// - Parameter name: Name to validate
    /// - Returns: Validation result with error if invalid
    @objc public func validateName(_ name: String) -> Result<Void, ValidationError> {
        if ValidationHelpers.isNameEmpty(name) {
            return .failure(.nameEmpty)
        }
        
        if ValidationHelpers.isNameTooShort(name) {
            return .failure(.nameTooShort)
        }
        
        if ValidationHelpers.isNameTooLong(name) {
            return .failure(.nameTooLong)
        }
        
        if !ValidationHelpers.hasValidNameCharacters(name) {
            return .failure(.nameInvalidCharacters)
        }
        
        return .success(())
    }
    
    /// Validate password confirmation
    /// - Parameters:
    ///   - password: Original password
    ///   - confirmPassword: Password confirmation
    /// - Returns: Validation result with error if invalid
    @objc public func validatePasswordConfirmation(_ password: String, confirmPassword: String) -> Result<Void, ValidationError> {
        if !ValidationHelpers.stringsMatch(password, confirmPassword) {
            return .failure(.passwordMismatch)
        }
        
        return .success(())
    }
    
    /// Validate complete registration credentials
    /// - Parameter credentials: Registration credentials to validate
    /// - Returns: Array of validation errors (empty if all valid)
    @objc public func validateRegistrationCredentials(_ credentials: RegistrationCredentials) -> [ValidationError] {
        var errors: [ValidationError] = []
        
        // Validate email
        switch validateEmail(credentials.email) {
        case .failure(let error):
            errors.append(error)
        case .success:
            break
        }
        
        // Validate password
        switch validatePassword(credentials.password) {
        case .failure(let error):
            errors.append(error)
        case .success:
            break
        }
        
        // Validate password confirmation
        switch validatePasswordConfirmation(credentials.password, confirmPassword: credentials.confirmPassword) {
        case .failure(let error):
            errors.append(error)
        case .success:
            break
        }
        
        // Validate name
        switch validateName(credentials.fullName) {
        case .failure(let error):
            errors.append(error)
        case .success:
            break
        }
        
        return errors
    }
    
    /// Validate complete login credentials
    /// - Parameter credentials: Login credentials to validate
    /// - Returns: Array of validation errors (empty if all valid)
    @objc public func validateLoginCredentials(_ credentials: UserCredentials) -> [ValidationError] {
        var errors: [ValidationError] = []
        
        // Validate email
        switch validateEmail(credentials.email) {
        case .failure(let error):
            errors.append(error)
        case .success:
            break
        }
        
        // Validate password (basic validation for login)
        if ValidationHelpers.isPasswordEmpty(credentials.password) {
            errors.append(.passwordEmpty)
        }
        
        return errors
    }
    
    // MARK: - Convenience Methods
    
    /// Validate email and return error message
    /// - Parameter email: Email address to validate
    /// - Returns: Error message if invalid, nil if valid
    @objc public func getEmailErrorMessage(_ email: String) -> String? {
        switch validateEmail(email) {
        case .failure(let error):
            return error.localizedDescription
        case .success:
            return nil
        }
    }
    
    /// Validate password and return error message
    /// - Parameter password: Password to validate
    /// - Returns: Error message if invalid, nil if valid
    @objc public func getPasswordErrorMessage(_ password: String) -> String? {
        switch validatePassword(password) {
        case .failure(let error):
            return error.localizedDescription
        case .success:
            return nil
        }
    }
    
    /// Validate name and return error message
    /// - Parameter name: Name to validate
    /// - Returns: Error message if invalid, nil if valid
    @objc public func getNameErrorMessage(_ name: String) -> String? {
        switch validateName(name) {
        case .failure(let error):
            return error.localizedDescription
        case .success:
            return nil
        }
    }
    
    /// Validate password confirmation and return error message
    /// - Parameters:
    ///   - password: Original password
    ///   - confirmPassword: Password confirmation
    /// - Returns: Error message if invalid, nil if valid
    @objc public func getPasswordConfirmationErrorMessage(_ password: String, confirmPassword: String) -> String? {
        switch validatePasswordConfirmation(password, confirmPassword: confirmPassword) {
        case .failure(let error):
            return error.localizedDescription
        case .success:
            return nil
        }
    }
}